### DISCUSSION TITLE
convert on a PDF produces highly pixelised version

### DISCUSSION THREAD

**Discussion by madduck** [ImageMagick/ImageMagick#5040] [Help] [answered]
I am running the following command on a 300×300 DPI PDF file in an attempt to remove the opaque white background:

```
convert -transparent white input.pdf output.pdf
```

Here is a bit of the input PDF:

![image](https://user-images.githubusercontent.com/195073/163014730-c587a59d-c196-493b-8dce-8b45dcc5000c.png)

and here is the same portion of the output PDF:

![image](https://user-images.githubusercontent.com/195073/163014972-967a9641-efd5-432b-9882-f8b83f3a64b6.png)

ImageMagick (6.9 here, still, unfortunately) has done a pretty bad job I'd say. I don't even need `-transparent white` to reproduce this, just converting PDF to PDF will break it like this.

I've tried `-depth 300x300` to match the DPI, but that made no difference. `-resolution` isn't a valid option.

What am I missing here? How can I get IM to not clobber the PDF while processing it?

**Comment by fmw42** [comment]
What version of ImageMagick?  What platform?

Post an example PDF.  

But to since IM rasterizes the PDF the way to increase quality is as follows:

`convert -density 288 input.pdf -resize 25% output.pdf`

This does a supersampling type process. It reads the input vector pdf at high resolution (4x) and resize down after rasterizing by 25% (1/4 x) to compensate for the increase in size due to the higher than normal (72 dpi) density of 288 specified. You can adjust the density and resize amount as desired.

For IM 7, replace convert with magick

**Reply by madduck** [reply]
I am using 6.9 on Linux.

I found I can reproduce this with any PDF, so I didn't post an example.

72/75 dpi isn't good enough. But when I run

```
convert -density $((300*4)) in.pdf -resize 25% out.pdf
```

I get:

```
convert-im6.q16: cache resources exhausted `/home/ssd/madduck/.tmp/magick-zrQEpFh3UYAItLFqx-JFt7iBsfamHyJJ1' @ error/cache.c/OpenPixelCache/4095.
```

**Reply by fmw42** [reply]
Your density (300*4 = 1200) is too large and the rasterized image has too many pixels and you run out of RAM before doing the -resize.  Also your result would be 16x file size for the resulting PDF.  If you set up your resources from the policy.xml file to use disk in place of RAM, you could do

`convert -density $((72*4*4)) in.pdf -resize $((100/16))% out.pdf`

assuming IM will allow those inline computations. Alternately,

`convert -density 1152 in.pdf -resize 6.25% out.pdf`

**Comment by snibgo** [comment, marked as answer]
```
convert -transparent white input.pdf output.pdf
```
That command would fail in IM v7 magick, because "-transparent" is an operation for images, but no images have been read when that operation is performed. Do the operation after reading the image:
```
convert input.pdf -transparent white output.pdf
```

> What am I missing here? How can I get IM to not clobber the PDF while processing it?

"-density" may help. If you don't show actual PDF files, we can only guess.

**Reply by madduck** [reply]
@snibgo @fmw42 Please find a sample PDF attached.
[sample.pdf](https://github.com/ImageMagick/ImageMagick/files/8478704/sample.pdf)

**Comment by snibgo** [comment]
Your sample.pdf contains one page, which contains one raster image and nothing else.

If you use IM to read that PDF, it will rasterize the page, which will result in resampling the embedded raster image, according to the "-density" setting. For example, "-density 360" before reading the PDF.

A better approach is simply to extract the embedded image. For that, I suggest the "pdfimages" program.
```
pdfimages -all sample.pdf abc
```
This creates the file abc-000.png. It has 2480x3506 pixels, and is grayscale, one bit/channel/pixel, black and white only.

Then you can do whatever processing you want on abc-000.png and, if you want, write that to a PDF file.

**Reply by madduck** [reply]
Thanks @snibgo. I was hoping that effectively, IM would work on the images, but it makes sense that it rasterises instead. I'll have to write a wrapper script to do the extraction and reassembly…
